File: tryton/tryton/gui/notification.py
# This file is part of Tryton.  The COPYRIGHT file at the top level of
# this repository contains the full copyright notices and license terms.

import gettext

from gi.repository import Gtk, Pango

import tryton.common as common
import tryton.rpc as rpc
from tryton.action import Action
from tryton.config import CONFIG
from tryton.gui.window import Window

_ = gettext.gettext


class NotificationMenu:

    def __init__(self, app):
        self.app = app
        self.button = Gtk.MenuButton()
        img = common.IconFactory.get_image(
            'tryton-notification', Gtk.IconSize.BUTTON)
        self.button.set_image(img)
        self.menu = Gtk.Menu()
        self.menu.connect('show', self.fill)
        self.button.set_popup(self.menu)

    def fill(self, menu):
        for child in menu.get_children():
            menu.remove(child)

        notifications = rpc.execute(
            'model', 'res.notification', 'get', rpc.CONTEXT)

        for notification in notifications:
            item = Gtk.MenuItem()

            hbox = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
            hbox.set_margin_top(4)
            hbox.set_margin_bottom(4)
            hbox.set_margin_start(6)
            hbox.set_margin_end(6)

            img = common.IconFactory.get_image(
                notification['icon'] or 'tryton-notification',
                Gtk.IconSize.MENU)
            hbox.pack_start(img, False, False, 0)

            vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2)

            label = Gtk.Label(notification['label'])
            label.set_xalign(0)
            vbox.pack_start(label, False, False, 0)

            description = Gtk.Label(notification['description'])
            description.set_xalign(0)
            description.get_style_context().add_class("dim-label")
            description.set_max_width_chars(30)
            description.set_line_wrap(True)
            description.set_line_wrap_mode(Pango.WrapMode.WORD_CHAR)
            description.set_ellipsize(Pango.EllipsizeMode.END)
            description.set_lines(2)
            vbox.pack_start(description, False, False, 0)

            hbox.pack_start(vbox, True, True, 0)
            item.add(hbox)

            item.connect('activate', self.open, notification)

            if notification['unread']:
                style = item.get_style_context()
                style.add_class('unread-notification')

            menu.append(item)

        if notifications:
            menu.append(Gtk.SeparatorMenuItem())

            def open_all_notifications(item):
                params = {
                    'domain': [['user', '=', rpc._USER]],
                    }
                Window.create('res.notification', **params)

            all_notifications = Gtk.MenuItem(label=_("All Notifications..."))
            all_notifications.connect('activate', open_all_notifications)
            menu.append(all_notifications)

        menu.show_all()
        self._update(0)

    def open(self, menuitem, notification):
        if notification.get('model') and notification.get('records'):
            params = {
                'domain': [['id', 'in', notification['records']]],
                }
            if len(notification['records']) == 1:
                params['res_id'] = notification['records'][0]
                params['mode'] = ['form', 'tree']
            Window.create(notification['model'], **params)
        if notification.get('action'):
            Action.execute(notification['action'])
        if notification['unread']:
            rpc.execute(
                'model', 'res.notification', 'mark_read',
                [notification['id']], rpc.CONTEXT)

    def _update(self, count):
        img = common.IconFactory.get_image(
            'tryton-notification', Gtk.IconSize.BUTTON,
            badge=2 if count else None)
        self.button.set_image(img)

    def notify(self, message):
        if message['type'] == 'user-notification':
            self._update(message['count'])
            for msg in message['content']:
                self.app.show_notification(CONFIG['client.title'], msg)

    def count(self):
        count = rpc.execute(
            'model', 'res.notification', 'get_count', rpc.CONTEXT)
        self._update(count)

File: sao/src/notification.js
/* This file is part of Tryton.  The COPYRIGHT file at the top level of
   this repository contains the full copyright notices and license terms. */

(function() {
    'use strict';

    class _NotificationMenu {
        constructor() {
            this.notify = this.notify.bind(this);
            this.el = jQuery('<ul/>', {
                'class': 'notification-menu dropdown-menu',
                'role': 'menu',
            });
            this.el.on('show.bs.dropdown', () => {
                this.fill();
                this.indicator.hide();
            });
            this.indicator = jQuery('<span/>', {
                'class': 'notification-badge',
            });
            let indicator_observer = new MutationObserver(() => {
                let indicators = jQuery('.notification-badge').not(this.indicator);
                indicators.text(this.indicator.text())
                indicators.toggle(this.indicator.css('display') !== 'none');
            });
            indicator_observer.observe(
                this.indicator.get(0), {
                    characterData: true,
                    attributes: true,
                    attributeFilter: ['style'],
                });
            this.indicator.hide();
        }

        fill() {
            let Notification = new Sao.Model('res.notification');
            Notification.execute('get', []).done((notifications) => {
                this.el.empty();
                for (let notification of notifications) {
                    let notification_item = jQuery('<div/>', {
                    }).append(jQuery('<span/>', {
                        'class': 'notification-label',
                        'text': notification.label,
                        'title': notification.label,
                    })).append(jQuery('<span/>', {
                        'class': 'notification-description',
                        'text': notification.description,
                        'title': notification.description,
                    }));
                    let link = jQuery('<a/>', {
                        'role': 'menuitem',
                        'href': '#',
                    }).click((evt) => {
                        evt.preventDefault();
                        this.open(notification)
                    }).append(notification_item);
                    let li = jQuery('<li/>', {
                        'class': 'notification-item',
                        'role': 'presentation',
                    });
                    let img = jQuery('<img/>', {
                        'class': 'icon',
                    });
                    link.prepend(img);
                    Sao.common.ICONFACTORY.get_icon_url(
                        notification.icon || 'tryton-notification')
                        .then(url => {
                            img.attr('src', url);
                            // Append only when the url is known to prevent
                            // layout shifts
                            li.append(link);
                        });
                    if (notification.unread) {
                        li.addClass('notification-unread');
                    }
                    this.el.append(li)
                }
                this.el.append(
                    jQuery('<li/>', {
                        'role': 'presentation',
                        'class': 'notification-item notification-action',
                    }).append(jQuery('<a/>', {
                        'role': 'menuitem',
                        'href': '#',
                        'title': Sao.i18n.gettext("All Notifications..."),
                    }).append(jQuery('<span/>', {
                        'class': 'caret',
                    })).click((evt) => {
                        evt.preventDefault();
                        let params = {
                            context: jQuery.extend({}, Sao.Session.current_session.context),
                            domain: [['user', '=', Sao.Session.current_session.user_id]],
                        };
                        params.model = 'res.notification';
                        Sao.Tab.create(params).done(() => {
                            this.indicator.hide();
                        });
                    }))
                );
                if (notifications.length > 0) {
                    this.el.append(
                        jQuery('<li/>', {
                            'role': 'separator',
                            'class': 'divider',
                        }));
                }
                let preferences_img = jQuery('<img/>', {
                    'class': 'icon',
                });
                let preferences = jQuery('<li/>', {
                    'class': 'notification-item',
                    'role': 'presentation',
                }).append(
                    jQuery('<a/>', {
                        'role': 'menuitem',
                        'href': '#',
                        'text': Sao.i18n.gettext("Preferences..."),
                    }).prepend(preferences_img
                    ).click((evt) => {
                        evt.preventDefault();
                        Sao.preferences();
                    }));
                Sao.common.ICONFACTORY.get_icon_url('tryton-launch')
                    .then(url => {
                        preferences_img.attr('src', url);
                    });
                this.el.append(preferences);
            });

        }

        open(notification) {
            let prms = [];
            if (notification.model && notification.records) {
                let params = {
                    context: jQuery.extend({}, Sao.Session.current_session.context),
                    domain: [['id', 'in', notification.records]],
                };
                if (notification.records.length == 1) {
                    params['res_id'] = notification.records[0];
                    params['mode'] = ['form', 'tree'];
                }
                params.model = notification.model;
                prms.push(Sao.Tab.create(params));
            }
            if (notification.action) {
                prms.push(Sao.Action.execute(notification.action));
            }
            jQuery.when.apply(jQuery, prms).done(() => {
                if (notification.unread) {
                    let Notification = new Sao.Model('res.notification');
                    Notification.execute('mark_read', [[notification.id]]);
                }
            });
        }

        _update(count) {
            if (count > 0) {
                if (count < 10) {
                    this.indicator.text(count);
                } else {
                    // Let's keep the text short
                    this.indicator.text('9+');
                }
                this.indicator.show();
            } else {
                this.indicator.hide();
            }
        }

        notify(message) {
            if (message.type == 'user-notification') {
                this._update(message.count);
                try {
                    if (Notification.permission == "granted") {
                        message.content.forEach((body) => {
                            new Notification(
                                Sao.config.title, {
                                    'body': body,
                                });
                        });
                    }
                } catch (e) {
                    Sao.Logger.error(e.message, e.stack);
                }
            }
        }

        count() {
            let Notification = new Sao.Model('res.notification');
            Notification.execute('get_count', [])
                .done((count) => {
                    this._update(count);
                });
        }
    }

    Sao.NotificationMenu = new _NotificationMenu();

}());
